ec2-cluster/modules/cluster-provisioner/main.tf

```diff
@@ -0,0 +1,79 @@
+resource "null_resource" "cluster_provisioner" {
+  count = var.provisioner_count
+  triggers = {
+    instance_ip = var.instance_ips[count.index]
+  }
+
+  connection {
+    host    = var.instance_ips[count.index]
+    type    = "ssh"
+    user    = var.ssh_user
+    timeout = "240s"
+  }
+
+  provisioner "file" {
+    content = templatefile("${path.module}/templates/join", {
+      genesis_ip          = var.instance_ips[0],
+      genesis_id          = var.instance_keys[0].id,
+      path_to_join_script = "/home/${var.ssh_user}/tessellation/join"
+    })
+    destination = "/home/${var.ssh_user}/tessellation/join-0"
+  }
+
+  provisioner "file" {
+    content = templatefile("${path.module}/templates/join", {
+      genesis_ip          = var.instance_ips[1],
+      genesis_id          = var.instance_keys[1].id,
+      path_to_join_script = "/home/${var.ssh_user}/tessellation/join"
+    })
+    destination = "/home/${var.ssh_user}/tessellation/join-1"
+  }
+
+  provisioner "file" {
+    content = templatefile("${path.module}/templates/join", {
+      genesis_ip          = var.instance_ips[2],
+      genesis_id          = var.instance_keys[2].id,
```

Comment:
1. Assumption that there are exactly 3 values in the `var.instance_ips`. What if there are 2 or 4? Please use `for_each` instead.
2. How about having a single file named `join` and another file called `hosts`. The `join` file would take an integer argument that would be a line number from the `hosts` file. This would result in calling `./join 1` instead of `./join-1`.